src/OpenCSGRenderer.cc

```diff
@@ -166,8 +193,247 @@ void OpenCSGRenderer::renderCSGProducts(const CSGProducts &products, GLint *shad
+}
+}
+
+void OpenCSGRenderer::draw(bool /*showfaces*/, bool showedges) const
+{
+	if (!getProductVertexSets().size()) {
+		GLint *shaderinfo = this->shaderinfo;
+		if (!shaderinfo[0]) shaderinfo = nullptr;
+
+		if (this->root_products) {
+				renderCSGProducts(*this->root_products, showedges ? shaderinfo : nullptr, false, false);
+		}
+		if (this->background_products) {
+				renderCSGProducts(*this->background_products, showedges ? shaderinfo : nullptr, false, true);
+		}
+		if (this->highlights_products) {
+				renderCSGProducts(*this->highlights_products, showedges ? shaderinfo : nullptr, true, false);
+		}
+	}
+
+	drawCSGProducts(showedges);
+}
+
+// Primitive for drawing using OpenCSG
+OpenCSGPrim *OpenCSGRenderer::createCSGPrimitive(const VertexSet &vertex_set, const GLuint vbo) const
+{
+  if (vertex_set.operation == OpenSCADOperator::INTERSECTION)
+    return new OpenCSGPrim(OpenCSG::Intersection, vertex_set.convexity, *this, vertex_set, vbo);
+
+  if (vertex_set.operation == OpenSCADOperator::DIFFERENCE)
+    return new OpenCSGPrim(OpenCSG::Subtraction, vertex_set.convexity, *this, vertex_set, vbo);
+
+  return nullptr;
+}
+
+void OpenCSGRenderer::renderCSGProducts(const CSGProducts &products, GLint * /*shaderinfo*/,
+										bool highlight_mode, bool background_mode) const
+{
+#ifdef ENABLE_OPENCSG
+	std::vector<Vertex> *render_buffer = new std::vector<Vertex>();
+	std::vector<VertexSet *> *vertex_sets = 0;
+	VertexSet *prev = 0;
+	VertexSet *vertex_set = 0;
+
+	GLuint vbo;
+
+	for(const auto &product : products.products) {
+		// one vbo for each product
+		if (product.intersections.size() || product.subtractions.size()) {
+			vertex_sets = new std::vector<VertexSet *>();
+
+			GLsizei combined_draw_size = 0;
+			Color4f last_color;
+			for(const auto &csgobj : product.intersections) {
+				if (csgobj.leaf->geom && vertex_sets) {
+					prev = (vertex_sets->empty() ? 0 : vertex_sets->back());
+					vertex_set = new VertexSet({true, OpenSCADOperator::INTERSECTION, csgobj.leaf->geom->getConvexity(), 0, 0, false, false});
+					GLintptr prev_start_offset = 0;
+					GLsizei prev_draw_size = 0;
+					if (prev) {
+						prev_start_offset = prev->start_offset;
+						prev_draw_size = prev->draw_size;
+					}
+
+					const Color4f &c = csgobj.leaf->color;
+					csgmode_e csgmode = get_csgmode(highlight_mode, background_mode);
+
+					ColorMode colormode = ColorMode::NONE;
+					if (highlight_mode) {
+						colormode = ColorMode::HIGHLIGHT;
+					} else if (background_mode) {
+						colormode = ColorMode::BACKGROUND;
+					} else {
+						colormode = ColorMode::MATERIAL;
+					}
+
+					Color4f c1; // = setColor(colormode, c.data(), shaderinfo);
+					if (getShaderColor(colormode, c, c1)) {
+						last_color = c1;
+					}
+
+					const_cast<OpenCSGRenderer *>(this)->create_surface(csgobj.leaf->geom, *render_buffer, *vertex_set, prev_start_offset, prev_draw_size,
+												 csgmode, csgobj.leaf->matrix, last_color);
```

Comment:
`VBORenderer::create_surface` can be made `const` (I checked) then the `const_cast` is unnecessary.